### DISCUSSION TITLE
HATs sharing UART

### DISCUSSION THREAD

**Discussion by gildas-declercq** [bareboat-necessities/lysmarine_gen#343] [Q&A] [answered]
Hi all, 

this is my first post I started my project with openplotter and I look now closely at BBN OS.

But my question is a bit off subjet.

I have pypilot hat, and RS485 hat, and other HATs that use UART0 . So I can't use them at the same time.
I understood that too lately.

I was thinking about having 2 rpi exchanging data via signalK,
another solution would be to connect the HAT on a separated breadboard, to connect the UART pins of the HAT on différents UART pins on the RPI.  this way I would use all the available UARTs on the RPI..

Is it a good or bad idea (I am still learning RPI)

thanks 
Gildas

**Comment by mgrouch** [comment]
Are you sure pypilot hat is using UART0?
Pypilot hat is a small screen and RF remote control. Pypilot hat is optional for pypilot. 
Pypilot motor controller is using UART gpio pins. 

For RS485 hat I think you can just find easy replacement with USB stick for RS485. 

What other hats do you have? Do they have usb options to connect them to pi?

**Reply by gildas-declercq** [reply]
you are right , it is the motor controller tthat use UART

I know I can find USB device for the NMEA2k access (in fact I will choose USB instead of hat device now when possible)
but I have this one and would like to use it

**Reply by gildas-declercq** [reply]
I have a gps hat, but can use it with usb, so I did

**Reply by mgrouch** [reply, marked as answer]
Too many hats will run into conflict. And hats are pi specific. You can’t use them with other devices. (Not without major hurdles)
Use USB where you can. 

For n2k you could use
if you want pypilot motor controller connected too

https://github.com/bareboat-necessities/bbn-nmea200-m5atom/tree/main/bbn-nmea2000-usb-gw-m5atom

**Reply by gildas-declercq** [reply]
motor controller and pypilot hat are bought to Sean. the controller is connected to the HAT, and if I understand correctly what Sean told me on openplotter forum, it is the motor controller that uses UART.

At the beginning I wanted to avoid USB, now I understand It should be the first choice .

more generally, do you think I can "tweak" a HAT as I said by using a breadboard and connect to one of the available UART ?

and thanks for your answers so far

**Reply by mgrouch** [reply]
Probably you can use UART to USB bridge and  effectively convert some of your hats into usb devices.

**Reply by gildas-declercq** [reply]
I will look at the solutions you suggest, especially https://github.com/bareboat-necessities/bbn-nmea200-m5atom/tree/main/bbn-nmea2000-usb-gw-m5atom looks intersting, better solution that the HAT.
thanx

**Reply by gildas-declercq** [reply]
Hi
one last more questions, would you suggest other intersting devices for my boat projet ?
I am looking for devices for water and fuel tank, motor temp, motor RPM, water detection... I would be intersted to see what you did yourself on your boat.
thanks

**Reply by mgrouch** [reply]
You can browse discussions here and find many links to other projects

https://github.com/bareboat-necessities/lysmarine_gen/discussions

also Getting Started Guide hardware section has many links

And here:
https://github.com/bareboat-necessities/bbn-m5stack-tough


If you looking for something specific you can always ask in BBN Discussions Forum:

https://github.com/bareboat-necessities/lysmarine_gen/discussions
